cryptography/hazmat/backends/openssl/backend.py

```diff
@@ -1055,6 +1063,77 @@ def _ec_key_set_public_key_affine_coordinates(self, ctx, x, y):
 
         return ctx
 
+    def generate_dh_parameters(self, generator, key_size):
+        if key_size < 512:
+            raise ValueError("DH key_size must be at least 512 bits")
+
+        dh_param_cdata = self._lib.DH_new()
+        assert dh_param_cdata != self._ffi.NULL
+        dh_param_cdata = self._ffi.gc(dh_param_cdata, self._lib.DH_free)
+
+        res = self._lib.DH_generate_parameters_ex(
+            dh_param_cdata,
+            key_size,
+            generator,
+            self._ffi.NULL
+        )
+        assert res == 1
+
+        return _DHParameters(self, dh_param_cdata)
+
+    def generate_dh_private_key(self, parameters):
+        dh_key_cdata = self._lib.DH_new()
+        assert dh_key_cdata != self._ffi.NULL
+        dh_key_cdata = self._ffi.gc(dh_key_cdata, self._lib.DH_free)
+
+        dh_key_cdata.p = self._lib.BN_dup(parameters._dh_cdata.p)
+        dh_key_cdata.g = self._lib.BN_dup(parameters._dh_cdata.g)
```

Comment:
This section could be simplified by using `DH *DHparams_dup(DH *);` to get a new `DH *` that already has the duplicated parameters. This was a macro prior to 1.0.0 so you'll want to put it in MACROS.